File: bindings/swift/Sources/WhatWG/Providers/UIProvider.swift
import Foundation

/// Protocol for providing UI functionality (alert, confirm, prompt).
///
/// Implement this protocol to provide modal dialog support.
///
public protocol UIProvider: AnyObject, Sendable {
    
    /// Shows an alert dialog.
    ///
    /// - Parameter message: The message to display.
    func alert(message: String) async
    
    /// Shows a confirmation dialog.
    ///
    /// - Parameter message: The message to display.
    /// - Returns: `true` if confirmed, `false` if cancelled.
    func confirm(message: String) async -> Bool
    
    /// Shows a prompt dialog.
    ///
    /// - Parameters:
    ///   - message: The message to display.
    ///   - defaultValue: The default input value.
    /// - Returns: The entered value, or `nil` if cancelled.
    func prompt(message: String, defaultValue: String?) async -> String?
    
    /// Opens a URL.
    ///
    /// - Parameters:
    ///   - url: The URL to open.
    ///   - target: The target window name.
    ///   - features: Window features.
    /// - Returns: A window handle, or `nil` if blocked.
    func open(url: URL?, target: String?, features: String?) async -> WindowHandle?
    
    /// Prints the current page.
    func print() async
    
    /// Scrolls the window.
    ///
    /// - Parameters:
    ///   - x: X offset.
    ///   - y: Y offset.
    func scrollTo(x: Double, y: Double)
    
    /// Scrolls the window by an offset.
    ///
    /// - Parameters:
    ///   - x: X delta.
    ///   - y: Y delta.
    func scrollBy(x: Double, y: Double)
}

/// A handle to a window.
public protocol WindowHandle: AnyObject, Sendable {
    /// Closes the window.
    func close()
    
    /// Focuses the window.
    func focus()
    
    /// Blurs the window.
    func blur()
    
    /// Posts a message to the window.
    func postMessage(_ message: Any, targetOrigin: String)
}

File: bindings/kotlin/whatwg/src/main/kotlin/com/whatwg/providers/UIProvider.kt
package com.whatwg.providers

/**
 * Interface for providing UI functionality (alert, confirm, prompt).
 *
 * Implement this interface to provide modal dialog support.
 */
interface UIProvider {

    /**
     * Shows an alert dialog.
     *
     * @param message The message to display.
     */
    suspend fun alert(message: String)

    /**
     * Shows a confirmation dialog.
     *
     * @param message The message to display.
     * @return `true` if confirmed, `false` if cancelled.
     */
    suspend fun confirm(message: String): Boolean

    /**
     * Shows a prompt dialog.
     *
     * @param message The message to display.
     * @param defaultValue The default input value.
     * @return The entered value, or `null` if cancelled.
     */
    suspend fun prompt(message: String, defaultValue: String? = null): String?

    /**
     * Opens a URL.
     *
     * @param url The URL to open.
     * @param target The target window name.
     * @param features Window features.
     * @return A window handle, or `null` if blocked.
     */
    suspend fun open(url: String?, target: String? = null, features: String? = null): WindowHandle?

    /**
     * Prints the current page.
     */
    suspend fun print()

    /**
     * Scrolls the window.
     *
     * @param x X offset.
     * @param y Y offset.
     */
    fun scrollTo(x: Double, y: Double)

    /**
     * Scrolls the window by an offset.
     *
     * @param x X delta.
     * @param y Y delta.
     */
    fun scrollBy(x: Double, y: Double)
}

/**
 * A handle to a window.
 */
interface WindowHandle {
    /**
     * Closes the window.
     */
    fun close()
    
    /**
     * Focuses the window.
     */
    fun focus()
    
    /**
     * Blurs the window.
     */
    fun blur()
    
    /**
     * Posts a message to the window.
     */
    fun postMessage(message: Any, targetOrigin: String)
}
